### PR TITLE
Added Filter inheritance hierarchy

### PR DESCRIPTION
**Pull request by lgfrbcsgo**
Refactored all filters. Filters now extend either the abstract `Filter` class directly or the abstract `IndexFilter` class which in turn extends the `Filter` class. So the inheritance structure looks like this:
```
Filter {abstract}
├── IndexFilter {abstract}
|     ├── PrefixTree
|     └── NaiveFilter
└── BloomFilter
```

The only difference between `Filter` and `IndexFilter` is that `IndexFilter` provides additional functionality to retrieve the indexes of matches within the text.

Also added Nocca JS to for testing. To run the tests for all filters, run `npm test`.

### CODE REVIEW

File: src/filters/prefix-tree.ts [hunk 1]
```diff
@@ -12,48 +15,86 @@
...
+    constructor(tokens? : string[]) {
+        super();
         this.root = new PTNode(); // not a leaf! we do not want to match any string!
-
-        if (terms) { // add provided terms if any
-            for (let term of terms) {
-                this.addTermToTree(term);
-            }
-        }
+        if (tokens) { super.addSearchTerms(tokens); }
```
Comment (nyxathid) [inline src/filters/prefix-tree.ts]: `super` will call the `addSearchTerms` in `Filter` 
-> `for (let token of tokens)` 
Does it break if `tokens` is `undefined`?

File: src/filters/prefix-tree.ts [hunk 2]
```diff
@@ -75,8 +116,9 @@ export interface PTElement {
      * Try to find any terms contained in the (sub-)tree structure in the search string at specified position.
      * @param searchStr  string to search
      * @param searchPos  position (index), position of the first character = 0
+     * @returns tuple of boolean (matches?), string (match) and number (match index)
      */
-    match(searchStr : string, searchPos : number) : boolean;
+    match(searchStr : string, searchPos : number) : [boolean, string, number];
```
Comment (nyxathid) [inline src/filters/prefix-tree.ts]: Nice!

File: src/filters/prefix-tree.ts (hunk 1 above)
Comment (lgfrbcsgo) [inline src/filters/prefix-tree.ts]: Probably. I've build this on the premise that the caller will check if it's undefined.

Comment (nyxathid) [inline src/filters/prefix-tree.ts]: Ah, yes, there is a check in the PrefixTree constructor. 👍

File: src/test-runs.ts [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
-import { BloomFilter } from "./bloom-filter";
+import { TermSearch } from "./term-search";
+import { BloomFilter } from "./filters/bloom-filter";
+import { Filter } from "./filters/filter";
+import { WetManager } from "./wet-manager";
+import { TermLoader } from "./utils/term-loader";
+import { PrefixTree } from "./filters/prefix-tree";
+import { IndexFilter } from "./filters/index-filter";
+import { Occurrence } from "./utils/occurrence";
 import {WetManager} from "./wet-manager";
```
Comment (nyxathid) [inline src/test-runs.ts]: duplicate import: WetManager

Comment (lgfrbcsgo) [inline src/test-runs.ts]: Merging on GitHub is hard... don't judge me...

**Pull request merged**